include/ocpp/v201/database_handler.hpp

```diff
@@ -178,16 +178,40 @@ class DatabaseHandler : public common::DatabaseHandlerCommon {
     /// charging profiles
 
     /// \brief Inserts or updates the given \p profile to CHARGING_PROFILES table
-    void insert_or_update_charging_profile(const int evse_id, const v201::ChargingProfile& profile);
+    void insert_or_update_charging_profile(
+        const int evse_id, const v201::ChargingProfile& profile,
+        const ChargingLimitSourceEnum charging_limit_source = ChargingLimitSourceEnum::CSO);
 
     /// \brief Deletes the profile with the given \p profile_id
-    void delete_charging_profile(const int profile_id);
+    bool delete_charging_profile(const int profile_id);
+
+    /// \brief Deletes the profiles with the given \p transaction_id
+    void delete_charging_profile_by_transaction_id(const std::string transaction_id);
 
     /// \brief Deletes all profiles from table CHARGING_PROFILES
-    void clear_charging_profiles();
+    bool clear_charging_profiles();
+
+    /// \brief Deletes all profiles from table CHARGING_PROFILES matching \p profile_id or \p criteria
+    bool clear_charging_profiles_matching_criteria(const std::optional<int32_t> profile_id,
+                                                   const std::optional<ClearChargingProfile>& criteria);
+
+    /// \brief Get all profiles from table CHARGING_PROFILES matching \p profile_id or \p criteria
+    std::vector<std::tuple<int, v201::ChargingProfile, ChargingLimitSourceEnum>>
+    get_charging_profiles_matching_criteria(const std::optional<int32_t> evse_id,
+                                            const ChargingProfileCriterion& criteria);
+
+    /// \brief Retrieves the charging profiles stored on \p evse_id
+    std::vector<v201::ChargingProfile> get_charging_profiles_for_evse(const int evse_id);
 
     /// \brief Retrieves all ChargingProfiles
+    std::vector<v201::ChargingProfile> get_all_charging_profiles();
+
+    /// \brief Retrieves all ChargingProfiles grouped by EVSE ID
     virtual std::map<int32_t, std::vector<v201::ChargingProfile>> get_all_charging_profiles_group_by_evse();
+
+    ChargingLimitSourceEnum get_charging_limit_source_for_profile(const int profile_id);
+
+    std::unique_ptr<common::SQLiteStatementInterface> new_statement(const std::string sql);
```

Comment:
Can this one be removed again? Does not seem to be used a lot if at all.

Reply:
We use this quite a bit currently within SmartChargingHandler. There's no easy way to remove it without having many very specific functions on the database handler. Long-term it would be better to use the repository pattern for the database handler, which would mean having a class that wraps the handler and provides operations that are specific to, for example, the charging profiles table.

Reply:
I haven't heard of the repository pattern before. If I google it it seems like this class is already kind of doing that, except for the fact that this does the operations for all tables. Instead using the repository pattern we would make a class for each table (or logical grouping thereof), is that correct?

I missed the use in SmartChargingHandler btw so this is fine to leave in for now.

Reply:
That would be correct, yes